src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -657,9 +479,359 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
+                return new Entry(originalEntry.State.WithAddedHostDocument(addAction.NewDocument, CreateTextAndVersionFunc(addAction.TextLoader)));
+
+            case RemoveDocumentAction removeAction:
+                return new Entry(originalEntry.State.WithRemovedHostDocument(removeAction.OriginalDocument));
+
+            case CloseDocumentAction closeAction:
+                {
+                    if (documentState is null)
+                    {
+                        throw new ArgumentNullException(nameof(documentState));
+                    }
+
+                    var state = originalEntry.State.WithChangedHostDocument(
+                        documentState.HostDocument,
+                        async () => await closeAction.TextLoader.LoadTextAndVersionAsync(LoadTextOptions, cancellationToken: default).ConfigureAwait(false));
+                    return new Entry(state);
+                }
+
+            case OpenDocumentAction openAction:
+                if (documentState is null)
+                {
+                    throw new ArgumentNullException(nameof(documentState));
+                }
+
+                {
+                    if (documentState.TryGetText(out var olderText) &&
+                        documentState.TryGetTextVersion(out var olderVersion))
+                    {
+                        var version = openAction.SourceText.ContentEquals(olderText) ? olderVersion : olderVersion.GetNewerVersion();
+                        var newState = originalEntry.State.WithChangedHostDocument(documentState.HostDocument, openAction.SourceText, version);
+                        return new Entry(newState);
+                    }
+                    else
+                    {
+                        var newState = originalEntry.State.WithChangedHostDocument(documentState.HostDocument, async () =>
+                        {
+                            olderText = await documentState.GetTextAsync().ConfigureAwait(false);
+                            olderVersion = await documentState.GetTextVersionAsync().ConfigureAwait(false);
+
+                            var version = openAction.SourceText.ContentEquals(olderText) ? olderVersion : olderVersion.GetNewerVersion();
+                            return TextAndVersion.Create(openAction.SourceText, version, documentState.HostDocument.FilePath);
+                        });
+
+                        return new Entry(newState);
+                    }
+                }
+
+            case DocumentTextLoaderChangedAction textLoaderChangedAction:
+                if (documentState is null)
+                {
+                    throw new ArgumentNullException(nameof(documentState));
+                }
+
+                return new Entry(originalEntry.State.WithChangedHostDocument(documentState.HostDocument, CreateTextAndVersionFunc(textLoaderChangedAction.TextLoader)));
+
+            case DocumentTextChangedAction textChangedAction:
+                if (documentState is null)
```

Comment:
Same comment as above

Reply:
FWIW, I suspect that you're probably OK using `AssumeNotNull` instead of throwing `ArgumentNullException`, which should clean up the code somewhat.